DragaliaAPI/Features/Event/EventDropService.cs

```diff
@@ -0,0 +1,440 @@
+                int emblem2Quantity = GenerateDropAmount(
+                    10d
+                        * (int)variation
+                        * (variation >= VariationTypes.Hell ? 5 * (int)variation : 1)
+                );
+                yield return new Entity(EntityTypes.RaidEventItem, itemId, emblem2Quantity);
+                break;
+            }
+        }
+
+        // Participating in the raid event will award Peregrine Blazons.
+        // TODO: No idea what the quantity should be for this lol
+        int blazonQuantity = GenerateDropAmount(50d * (int)variation);
+        yield return new Entity(
+            EntityTypes.RaidEventItem,
+            itemDict[RaidEventItemType.SummonPoint],
+            blazonQuantity
+        );
+    }
+
+    private IEnumerable<Entity> ProcessCombatEventDrops(
+        QuestData quest,
+        EventData evt,
+        PlayRecord record
+    )
+    {
+        // https://dragalialost.wiki/w/Defensive_Events
+
+        Dictionary<CombatEventItemType, int> itemDict = MasterAsset.CombatEventItem.Enumerable
+            .Where(x => x.EventId == evt.Id)
+            .ToDictionary(x => x.EventItemType, x => x.Id);
+
+        VariationTypes variation = quest.VariationType;
+
+        // General versions of Defensive Events have only one form of currency, Primal Crystals,
+        // which can be exchanged in the Event Shop for various rewards.
+        int primalCrystalQuantity = GenerateDropAmount(10d * (int)variation);
+        yield return new Entity(
+            EntityTypes.CombatEventItem,
+            itemDict[CombatEventItemType.ExchangeItem],
+            primalCrystalQuantity
+        );
+
+        // In general versions of defensive events, Stratagems are a resource that can be used to challenge EX defensive battles.
+        // They can be obtained as drops from Master-difficulty quests.
+        if (variation == VariationTypes.VeryHard)
+        {
+            int stratagemQuantity = GenerateDropAmount(2d);
+            yield return new Entity(
+                EntityTypes.CombatEventItem,
+                itemDict[CombatEventItemType.AdventItem],
+                stratagemQuantity
+            );
+        }
+    }
+
+    private IEnumerable<Entity> ProcessClb01EventDrops(
+        QuestData quest,
+        EventData evt,
+        PlayRecord record
+    )
+    {
+        // Also https://dragalialost.wiki/w/Defensive_Events
+
+        Dictionary<Clb01EventItemType, int> itemDict = MasterAsset.Clb01EventItem.Enumerable
+            .Where(x => x.EventId == evt.Id)
+            .ToDictionary(x => x.EventItemType, x => x.Id);
+
+        DungeonTypes type = quest.DungeonType;
+        VariationTypes variation = quest.VariationType;
+
+        // T2 currency can only be obtained from event quests made available in the second half of the event
+        // TODO: I have no idea how to check this, hopefully DungeonType should be good enough
```

Comment:
what lead you to use dungeontype?

Probably easier to just reward it from all quests even if not strictly 'authentic'

Reply:
looking at the masterdata for all quests of these events there is a switch in the middle of them where the dungeontype switches, so I thought that might be the first half / second half divide

Reply:
the clb02 event featured some normal quests / boss battles and some defensive battles, could be that? Will look into it